glusterd2/commands/volumes/brick-replace.go

```diff
@@ -0,0 +1,295 @@
+				brickIndex = i
+				srcBrickInfo = brick
+			}
+			// Avoid duplicate appends
+			if _, ok := m[brick.PeerID.String()]; !ok {
+				m[brick.PeerID.String()] = true
+				peerIDs = append(peerIDs, brick.PeerID)
+			}
+		}
+	}
+
+	/*if srcBrickInfo == (brick.Brickinfo{}) {
+		//Brick not found, quit with error msg
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "source brick not found")
+		return
+	}*/
+
+	allPeers, err := peer.GetPeers()
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+
+	zones := make(map[string]bool)
+	// Get all zones of bricks in volume
+	for _, peerID := range peerIDs {
+		peer, err := peer.GetPeer(peerID.String())
+		if err != nil {
+			restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+			return
+		}
+		if _, ok := zones[peer.Metadata["_zone"]]; !ok {
+			zones[peer.Metadata["_zone"]] = true
+		}
+	}
+
+	// Find peers which has zone apart from the ones already being used in volume.
+	var validPeers []*peer.Peer
+	for _, p := range allPeers {
+		if _, ok := m[p.ID.String()]; !ok {
+			if _, valid := zones[p.Metadata["_zone"]]; !valid {
+				validPeers = append(validPeers, p)
+			}
+		}
+	}
+
+	// Get available VG
+	availableVgs, err := getAvailableVgs(validPeers)
+	if err != nil {
+		return
+	}
+
+	// TODO: check for available vgs in zones already being used in volume.
+	if len(availableVgs) == 0 {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "No volume groups are available")
+		return
+	}
+
+	mtabEntries, err := volume.GetMounts()
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	// Get source brick information like size etc
+	brickInfo, err := volume.BrickStatus(srcBrickInfo, mtabEntries)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	// Get new brick from the available vgs
+	newBrick := getNewBrick(availableVgs, brickInfo, vol, brickCount)
+
+	peerID := uuid.Parse(newBrick.PeerID)
+	if peerID == nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "peer id of new brick could not be parsed")
+		return
+	}
+
+	nodes := []uuid.UUID{peerID}
+	txn, err := transaction.NewTxnWithLocks(ctx, volname)
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+	defer txn.Done()
+
+	txn.Steps = []*transaction.Step{
+		{
+			DoFunc: "brick-replace.txnPrepareBrick",
+			Nodes:  nodes,
+		},
+		{
+			DoFunc: "brick-replace.ReplaceVolinfo",
+			Nodes:  []uuid.UUID{gdctx.MyUUID},
+		},
+		{
+			DoFunc:   "vol-create.InitBricks",
+			UndoFunc: "vol-create.UndoInitBricks",
+			Nodes:    nodes,
+		},
+		{
+			DoFunc: "brick-replace.StartBrick",
+			Nodes:  nodes,
+		},
+		{
+			DoFunc:   "vol-create.StoreVolume",
+			UndoFunc: "vol-create.UndoStoreVolume",
+			Nodes:    []uuid.UUID{gdctx.MyUUID},
+		},
```

Comment:
Notify is required to notify all clients about new brick

Reply:
Done